lib/DxilDia/DxilDiaSymbolManager.cpp

```diff
@@ -1529,13 +1563,30 @@ HRESULT dxil_dia::hlsl_symbols::SymbolManagerInit::CreateCompositeType(DWORD dwP
   TypeInfo *udtTI;
   IFR(GetTypeInfo(CT, &udtTI));
   auto udtScope = BeginUDTScope(udtTI);
-  for (llvm::DINode *N : CT->getElements()) {
-    if (auto *Field = llvm::dyn_cast<llvm::DIType>(N)) {
-      DWORD dwUnusedFieldID;
-      IFR(CreateType(Field, &dwUnusedFieldID));
+  if (CT->getElements().size() == 0) {
+    // "Resources" (textures, samplers, etc.) are composite types without any elements,
+    // but they do have a size.
+    udtTI->AppendSize(CT->getSizeInBits());
+  } else {
+    for (llvm::DINode *N : CT->getElements()) {
+      if (auto *Field = llvm::dyn_cast<llvm::DIType>(N)) {
+        DWORD dwUnusedFieldID;
+        if (Field->getName() == "MaterialTextureBilinearWrapedSampler") {
+          Field->getName();
+        }
+        IFR(CreateType(Field, &dwUnusedFieldID));
+        if (Field->getTag() == llvm::dwarf::DW_TAG_inheritance) {
+          // The base class is a type of its own, so will have contributed to
+          // its own TypeInfo. But we still need to remember the size that it
+          // contributed to this type:
+          auto *DerivedType = llvm::dyn_cast<llvm::DIDerivedType>(Field);
```

Comment:
If we expect this cast to work, we should just use llvm::cast. This way if it fails it'll be a llvm fatal error instead of an AV.

Reply:
Looks like the base class always has to be DIDerivedType, so yes that should always work. Changed.